### PR TITLE
Implement instances for monad type class hierarchy

### LINKED ISSUE #21: Implement Monad type class hierarchy instances for Indexed
Per my [exploration on the indexed capability design pattern (or indexed finally tagless pattern?)](https://discourse.purescript.org/t/exploration-indexed-capability-design-pattern-with-indexed-monad-transformers/956/2), a default implementation (i.e. `IxExceptT`) for `IxMonadError` cannot be defined because `ibind` always assumes that some forward progress is made. However, if we treat `Indexed` as a monad whose indices don't change when normal `bind` is used, we can then use regular monad transformers without issue.

So I propose adding such instances.

### PR DESCRIPTION
**Pull request by JordanMartinez**
Fixes #21

### CODE REVIEW

File: src/Data/Indexed.purs [hunk 1]
```diff
@@ -20,6 +20,23 @@ derive newtype instance ord1Indexed ∷ Ord1 m ⇒ Ord1 (Indexed m i o)
...
+  pure = Indexed <<< pure
+
+instance bindIndexed :: Bind m ⇒ Bind (Indexed m x x) where
+  bind (Indexed ma) f = Indexed do
+    a <- ma
+    let (Indexed mb) = f a
+    mb
+
+instance monadIndexed :: Monad m ⇒ Monad (Indexed m x x)
+
```
Comment (garyb) [inline src/Data/Indexed.purs:28]: You could use newtype deriving for these instead: 

```purescript
derive newtype instance functorIndexed :: Functor m ⇒ Functor (Indexed m x x)
derive newtype instance applyIndexed :: Apply m ⇒ Apply (Indexed m x x)
derive newtype instance applicativeIndexed :: Applicative m ⇒ Applicative (Indexed m x x)
derive newtype instance bindIndexed :: Bind m ⇒ Bind (Indexed m x x)
derive newtype instance monadIndexed :: Monad m ⇒ Monad (Indexed m x x)
```

**Comment by JordanMartinez** [comment]
Wow... Silly me... I'll fix that later on today.

**Comment by JordanMartinez** [comment]
Fixed

**Comment by srghma** [comment]
awesome, was about to implement this myself

@garyb , friendly ping)

**Comment by garyb** [comment]
Thanks! Sorry about the ridiculous delay.

**Comment by JordanMartinez** [comment]
:smile: No worries!

**Pull request merged**